File: unittests/nvqpp/backends/quake_backend/test_extern_call.py
# ============================================================================ #
# Copyright (c) 2026 NVIDIA Corporation & Affiliates.                          #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #
import cudaq
import numpy as np
import sys

cudaq.set_target("quake_fake")


@cudaq.extern_kernel
def wait(duration: float, q: cudaq.qubit) -> None:
    ...


@cudaq.kernel
def ramsey_single(wait_duration: float) -> bool:
    qubit = cudaq.qubit()
    rx(np.pi / 2, qubit)
    wait(wait_duration, qubit)
    rx(np.pi / 2, qubit)
    return mz(qubit)


try:
    results = cudaq.run(ramsey_single, 1.0, shots_count=100)
    assert len(results) == 100, f"expected 100 results, got {len(results)}"
    for shot in results:
        assert shot in (True, False), f"unexpected result {shot}"
except Exception as e:
    print(e)
    sys.exit(1)

File: unittests/nvqpp/backends/quake_backend/test_extern_call.cpp
/*******************************************************************************
 * Copyright (c) 2026 NVIDIA Corporation & Affiliates.                         *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

#include "qm_wait.h"

#include <cudaq.h>
#include <iostream>

// A Ramsey sequence. The mock server checks for the call in wire form.
__qpu__ bool ramsey_single(double wait_duration) {
  cudaq::qubit q;
  rx(M_PI_2, q);
  __qm__wait_function(wait_duration, q);
  rx(M_PI_2, q);
  return mz(q);
}

int main() {
  auto results = cudaq::run(100, ramsey_single, 1.0);
  if (results.size() != 100) {
    std::cerr << "ramsey_single returned " << results.size()
              << " results, expected 100\n";
    return 1;
  }
  std::cout << "Successfully sampled a kernel with an external quantum call\n";
  return 0;
}
